main/exercise/ptest_category.class.php

```diff
@@ -0,0 +1,1174 @@
+     *
+     * @return string
+     */
+    public function returnCategoryForm(Exercise $exercise)
+    {
+        $categories = $this->getListOfCategoriesForTest($exercise);
+        $saved_categories = $exercise->getCategoriesInExercise();
+        $return = null;
+
+        if (!empty($categories)) {
+            $nbQuestionsTotal = $exercise->getNumberQuestionExerciseCategory();
+            $exercise->setCategoriesGrouping(true);
+            $real_question_count = count($exercise->getQuestionList());
+
+            $warning = null;
+            if ($nbQuestionsTotal != $real_question_count) {
+                $warning = Display::return_message(
+                    get_lang('CheckThatYouHaveEnoughQuestionsInYourCategories'),
+                    'warning'
+                );
+            }
+
+            $return .= $warning;
+            $return .= '<table class="data_table">';
+            $return .= '<tr>';
+            $return .= '<th height="24">'.get_lang('Categories').'</th>';
+            $return .= '<th width="70" height="24">'.get_lang('Number').'</th></tr>';
+
+            $emptyCategory = [
+                'id' => '0',
+                'name' => get_lang('NoCategory'),
+                'description' => '',
+                'iid' => '0',
+                'title' => get_lang('NoCategory'),
+            ];
+
+            $categories[] = $emptyCategory;
+
+            foreach ($categories as $category) {
+                $cat_id = $category['iid'];
+                $return .= '<tr>';
+                $return .= '<td>';
+                $return .= Display::div($category['name']);
+                $return .= '</td>';
+                $return .= '<td>';
+                $value = isset($saved_categories) && isset($saved_categories[$cat_id]) ? $saved_categories[$cat_id]['count_questions'] : -1;
+                $return .= '<input name="category['.$cat_id.']" value="'.$value.'" />';
+                $return .= '</td>';
+                $return .= '</tr>';
+            }
+
+            $return .= '</table>';
+            $return .= get_lang('ZeroMeansNoQuestionWillBeSelectedMinusOneMeansThatAllQuestionsWillBeSelected');
+        }
+
+        return $return;
+    }
+
+    /**
+     * Return true if a category already exists with the same name.
+     *
+     * @param string $name
+     * @param int    $courseId
+     *
+     * @return bool
+     */
+    public static function categoryTitleExists($name, $courseId = 0)
+    {
+        $categories = self::getCategoryListInfo('title', $courseId);
+        foreach ($categories as $title) {
+            if ($title == $name) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Return the id of the test category with title = $in_title.
+     *
+     * @param string $title
+     * @param int    $courseId
+     *
+     * @return int is id of test category
+     */
+    public static function get_category_id_for_title($title, $courseId = 0)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The variable $out_res is not named in camelCase.](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5018478)